Bringing support for all pre-fragment operations and some special fun...

GlenDC opened the pull request:
Bringing support for all pre-fragment operations and some special functions.
This does break the API (`webgl` function).

#### References

The main reference has been the khornos manual pages and [the WebGL API 1.0 reference cheat sheet](https://www.khronos.org/files/webgl/webgl-reference-card-1_0.pdf).

#### Description

This update allows users to enable/disable the capabilities they want and use the pre-fragmented operations available. This also removes the default call to enable DepthTests, as this should be defined by the user if she wants it.

#### Background

First I wanted to go really minimalistic to give all the freedom the user wants. But this was quit error prone, non-type safe and gives no protection at all at compile time. Therefore I decided to go for what you can find in this pull request. It does expand the code quite a bit. However it allows for a well specified and type-safe way to use these functions.

Hopefully we find an agreement on this so that me and others can start this functionality in our projects that depend on this great library. Also as I didn't do so yet, thank you for all the work that has been put in this library so far, as it really makes using WebGL fun and surprisngly easy.

####  Example webgl call

```elm
import WebGL as GL

{-| ... -}

GL.webgl (w, h)
  [ GL.Enable GL.Blend
  , GL.Disable GL.DepthTest
  , GL.BlendEquation GL.Add
  , GL.BlendFunc (GL.SrcAlpha, GL.OneMinusSrcAlpha)
  ]
  [ GL.entity vertexShader fragmentShader mesh { mat = matrix } ]
```

hughfdjackson commented:
@GlenDC This looks great!  I do wonder if it'd be worth introducing an additional function - `webglWithOptions` or similar - instead of needing a new major version? 

I don't know what the general feel about this is in the elm community.

GlenDC commented:
Thanks @hughfdjackson  :)

Having it as another option is certainly an option, but I don't think it's a good idea. As I think it's quite rare that you want to rely on the default values for everything. And if so, `[]` can always be passed, similar to how other libraries do it, such as `Elm.Html`

Also I think it's a big enough change, that forcing a new major version is probably not a bad idea. Would love to hear the feedback of the maintainers on this, such as @johnpmayer, @evancz, @deadfoxygrandpa.

Do you agree with this approach? Any functionality missing related to this change? Any functionality that's too much and should probably not be there?

hughfdjackson commented:
@GlenDC That makes sense :) While this code is in pre-review limbo, I'll be using your branch locally!

nphollon commented:
I was about to request something like this!
What would you think about adding support for getExtension()? I would like to enable [EXT_frag_depth](https://www.khronos.org/registry/webgl/extensions/EXT_frag_depth/), but this needs to be done before the shaders are compiled

deadfoxygrandpa commented:
I like this idea and it seems like a pretty easy way to use it

GlenDC commented:
@nphollon I can take a look at it over the following days to see if I can come up with a solution that does't make the API convoluted. If you already had something in mind, feel free to share.

nphollon commented:
@GlenDC I looked into it a little more, and I think it's safe to say that my suggestion is out of scope for this pull request :-) Since getExtension() can return an arbitrary JavaScript object, I imagine extension support will probably have to be done on a case-by-case basis.

But to get back on-topic, I like what you've done, and I think @hughfdjackson suggestion of introducing a new function is prudent.

GlenDC commented:
@nphollon I've looked into it as well, and it made me remember that I concidered similar functionality. Back then I also concluded that it would be to much. I think it's indeed out of scope for this PR.

Not sure who is mainly maintaining this repository these days, but it would be nice to see some progress on this, as I need this kind of functionality fairly soon.

evancz commented:
I'm not *really* a maintainer on this, I happened to get added as a contributor though. I glanced through, and the big thing I'd say is: do not expose the constructors of a union type directly!

Do you want people to be able to pattern match on it? Do you want it to be a major change when you add a new item to that list? What are the chances that something is added?

Just looking through, it also seems like a kind of low level way to introduce all this. Can this be done without exposing a bunch of functions?

Like I said, I am not a maintainer, I am asking these questions as a stop gap, I can't focus on this repo!

GlenDC commented:
Thanks @evancz, I appreciate your comments & tips.And thanks for th time! :+1: 
Very critical insighs, going to give that some time to sink.
It is indeed very low level, but at the same time we do need some form of control, so I suppose having a way to expand in the future without having to change major versions all the time, is probably a good thing.

johnpmayer commented:
Two main things

1) 

Would rather support keeping webgl with implicit defaults

``` Elm
defaultConfiguration : FragmentConfig
webglWithConfig : FragmentConfig -> (Int,Int) -> List Entity -> Element
webgl = webglWithConfig defaultConfiguration 
```

Second, and this I'm asking becuase I don't actually know, but do any of these options make more sense to configure an Entity rather than the whole gl context?

GlenDC commented:
@johnpmayer I have no strong feelings about the seperation of function, I said my opinion on it before, but I can live with the seperation of config-vs-default functions.

It does make sense to configure them for each context, as you might have several contexts that require different settings, and in the case that you have also just one context, it doesn't matter either.

GlenDC commented:
@johnpmayer I added a commit that makes webgl use implicit configs once again. I also added functionality to enable configuration on a per-entity basis.

Please let me know what you think about this latest change.

GlenDC commented:
@johnpmayer is there actually something you still want me to do?
If you're happy with the latest state, could we please move on?

johnpmayer commented:
Sorry. I've been hung up on adding some sort of test examples, but haven't figured out a good story. Given the state of local dependencies (not existing) and the time I have to spend, I think it makes more sense to just accept the API additions and then do bug fixes as they're reported.
